sysml.library/Kernel Library/VectorFunctions.kerml

```diff
@@ -0,0 +1,192 @@
+		inv zeroAddition { w == null or isZeroVector(w) implies u == w }
+		inv commutivity { w != null implies u == w + v }
+	}
+	
+	/**
+	 * With two arguments, returns the difference of two VectorValues. With one arguments, returns the inverse
+	 * of the given VectorValue, that is, the VectorValue that, when added to the original VectorValue, results in
+	 * the zeroVector.
+	 */
+	abstract function '-' specializes DataFunctions::'-' (v: VectorValue[1], w: VectorValue[0..1]) u: VectorValue[1] {
+		inv negation { w == null implies isZeroVector(v + u) }
+		inv difference { w != null implies v + u == w }
+	}
+	
+	/**
+	 * Return the sum of a collection of VectorValues. If the collection is empty, return a given zero vector.
+	 */
+	abstract function sum0 (coll: VectorValue[*] nonunique, zero: VectorValue[1]) s: VectorValue[1] {
+		inv precondition { isZeroVector(zero) }
+		coll->reduce '+' ?? zero
+	}
+
+	/* Functions specific to NumericalVectorValues. */
+	
+	/**
+	 * Construct a NumericalVectorValue whose elements are a non-empty list of component NumericalValues.
+	 * The dimension of the NumericalVectorValue is equal to the number of components.
+	 */
+	function VectorOf(components: NumericalValue[1..*] ordered nonunique): NumericalVectorValue {
+		NumericalVectorValue (
+			dimension = size(components),
+			elements = components
+		)
+	}
+	
+	/**
+	 * Scalar product of a NumericalValue and a NumericalVectorValue.
+	 */
+	abstract function ScalarVectorMult specializes DataFunctions::'*' (x: NumericalValue[1], v: NumericalVectorValue[1]) w: NumericalVectorValue[1] {
+		inv scaling { Length(w) == x * Length(v) }
+		inv zeroLength { isZeroVector(w) implies isZero(Length(w))}
+	}
+	alias '*' for ScalarVectorMult;
+	
+	/**
+	 * Scalar product of a NumericalVectorValue and a NumericalValue, which has the same value as the scalar product of the
+	 * NumericalValue and the NumericalVectorValue.
+	 */
+	abstract function VectorScalarMult specializes DataFunctions::'*' (v: NumericalVectorValue[1], x: NumericalValue[1]): NumericalVectorValue[1] {
+		ScalarVectorMult(x, v)
+	}
+	
+	/**
+	 * Scalar quotient of a NumericalVectorValue and a NumericalValue, defined as the scalar product of the inverse of the 
+	 * NumericalValue and the NumericalVectorValue.
+	 */
+	abstract function VectorScalarDiv specializes DataFunctions::'/' (v: NumericalVectorValue[1], x: NumericalValue[1]): NumericalVectorValue[1] {
+		ScalarVectorMult(1.0 / x, v)
+	}
+
+	/**
+	 * Inner product of two NumericalVectorValues.
+	 */
+	abstract function Inner specializes DataFunctions::'*'(v: NumericalVectorValue[1], w: NumericalVectorValue[1]) x: NumericalValue[1] {
+		inv commmutivity { x == Inner(w, v) }
+		inv zeroInner { isZeroVector(v) or isZeroVector(w) implies isZero(x)}
+	}
+	
+	/**
+	 * The length of a NumericalVectorValue, as a NumericalValue.
+	 */
+	abstract function Length(v: NumericalVectorValue[1]) l : NumericalValue[1] {
```

Comment:
Should we use Magnitude instead of Length? In general a NumericalVectorValue may represent any kind of physical quantity.

Reply:
Good point, though `Magnitude` is a bit long. An alternative would be `norm`. Do you think that would be familiar enough to people?

Reply:
Hey, Julia uses `norm` (https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.norm), what can I say? So good for me, clear and short.

Reply:
Updated in commit [3505da2](https://github.com/Systems-Modeling/SysML-v2-Pilot-Implementation/pull/337/commits/3505da240c3983885d275d65d636fde6adcf7ad6).